### DISCUSSION TITLE
Too many redirect daphne, nginx,django channel ubuntu 20

### DISCUSSION THREAD

**Discussion by halhwadi** [django/daphne#401] [Q&A] [answered]
My site is working properly but the issue right now is redirecting http to https, I tried to configure the nginx but every time I get "Too many redirect" eorr, i have clear the cache but with no luck

when I add **return 301 https://$server_name$request_uri;** or rewrite command to server block  I got Too many redirect error, I tried to comment this line **proxy_pass http://channels-backend** when adding the redirect but does not help...

daphne.sock file 

```
[Unit]
Description=daphne daemon
Requires=daphne.socket
After=network.target

[Service]
Type=simple
WorkingDirectory=/home/husam/bikum/
ExecStart=daphne -b 0.0.0.0 -p 8000 --proxy-headers Bikum.asgi:application

[Install]
WantedBy=multi-user.target
```

nginx confg:
```
upstream channels-backend {
    server localhost:8000;
}

server {
    listen 80;
    server_name bikum.org www.bikum.org;

  location = /favicon.ico { access_log off; log_not_found off; }
    location /static/ {
        root /home/husam/bikum;
   }
  
  location /media/ {
      root /home/husam/bikum;
     }

     location / {
        try_files $uri @proxy_to_app;
    }
     
    location @proxy_to_app {
        proxy_pass http://channels-backend;

        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        proxy_redirect off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $server_name;
    }

    listen 443 ssl;
    ssl_certificate /etc/letsencrypt/live/bikum.org/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/bikum.org/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
   
}

```

**Comment by carltongibson** [comment]
> when I add return 301 https://$server_name$request_uri; or rewrite command to server block I got Too many redirect error

You want the redirect only if the protocol is `http`. This is an NGINX question. I'd try Server Fault for help.

**Comment by halhwadi** [comment, marked as answer]
Thanks for your reply and assistance
Yes it has nothing with Daphne..

Issue has been solved by:
1- adding protocol header to nginx:
proxy_set_header X-Forwarded-Proto $scheme;

2- Adding the following lines to setting.py file:
SECURE_SSL_REDIRECT = True
SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')

Third point is adding --proxy-headers to daphne run command in sock file; but not sure if it has any affect on sharing protocol with Django 

Best Regards

**Reply by carltongibson** [reply]
Super. Glad you worked it out.

**Reply by AKRASH-Nadeem** [reply]
it is not working for me when i used --proxy-headers it is not passing x-forwarded-proto.
my django application is getting http instead of https in x-forwarded-proto

**Reply by AKRASH-Nadeem** [reply]
> it is not working for me when i used --proxy-headers it is not passing x-forwarded-proto. my django application is getting http instead of https in x-forwarded-proto

i have found the solution of my problem.

The problem was with cloudflare. i had proxied enabled in cloudflare because of which i was getting request from an ip instead of a domain and wrong nginx config was being used.

I just disabled proxied in CloudFlare DNS record and it worked.
